StreamChatClient_v3/ChatClient.swift

```diff
@@ -151,80 +142,264 @@ public class Client<ExtraData: ExtraDataTypes> {
+    /// if you want to keep receiving updates about the newly set user.
+    ///
+    /// - Parameters:
+    ///   - userId: The id of the new current user.
+    ///   - userExtraData: You can optionally provide additional data to be set for the user. This is an equivalent of
+    ///   setting the current user detail data manually using `CurrentUserController`.
+    ///   - token: You can provide a token which is used for user authentication. If the `token` is not explicitly provided,
+    ///   the client uses `ChatClientConfig.tokenProvider` to obtain a token. If you haven't specified the token provider,
+    ///   providing a token explicitly is required. In case both the `token` and `ChatClientConfig.tokenProvider` is specified,
+    ///   the `token` value is used.
+    ///   - completion: Called when the new user is successfully set.
+    public func setUser(userId: UserId,
+                        userExtraData: ExtraData.User? = nil,
+                        token: Token? = nil,
+                        completion: ((Error?) -> Void)? = nil) {
+        guard token != nil || config.tokenProvider != nil else {
+            log.assert(false,
+                       "The provided token is `nil` and `ChatClientConfig.tokenProvider` is also `nil`. You must either provide " +
+                           "a token explicitly or set `TokenProvider` in `ChatClientConfig`.")
+            completion?(ClientError.MissingToken())
+            return
+        }
+        
+        guard userId != currentUserId else {
+            log.warning("New user with id:<\(userId)> is not set because it's similar to the already logged-in user.")
+            completion?(nil)
+            return
+        }
+        
+        disconnect()
+        
+        prepareEnvironmentForNewUser(userId: userId, role: .user, extraData: userExtraData) { error in
+            guard error == nil else {
+                completion?(error)
+                return
+            }
+            
+            if let token = token {
+                self.currentToken = token
+                self.connect(completion: completion)
+                
+            } else {
+                // Use `tokenProvider` to get the token
+                self.refreshToken { [weak self] (error) in
+                    guard error != nil else {
+                        completion?(error)
+                        return
+                    }
+                    
+                    self?.connect(completion: completion)
+                }
+            }
+        }
+    }
+    
+    /// Connects `Client` to the chat servers. When the connection is established, `Client` starts receiving chat updates.
+    ///
+    /// - Parameter completion: Called when the connection is established. If the connection fails, the completion is
+    /// called with an error.
+    public func connect(completion: ((Error?) -> Void)? = nil) {
```

Comment:
This is shouldn't be public. 
We should only connect on `setUser`, `setAnonymousUser` and etc

Reply:
When developers set `setUser` then after that we should keep connection and reconnect if went from background. 
We should stop do that if developers call `disconnect()` deliberately.

Reply:
If we have `disconnect()` public, we need to have `connect()` public, too. Is there any reason why to *not* expose this API?

Reply:
Because we have to connect on `setUser` and keep connection as much as possible.
`disconnect` basically will logout the user.

Reply:
I'm for giving users more control, login/logout is restrictred. I prefer login/connect/disconnect/logout as it's more granular

Reply:
We discussed that in the past. Connect should happened on setUser.

Reply:
It does. Please check the implementation, or the flow described in the PR description. `connect` is called on `setUser` automatically. There's no reason, why users shouldn't be able to call `disconnect()` and `connect()` if they want :) Especially, when we have this functionality implemented.